Feat: Support arrays_overlap function

erenavsarogullari opened the pull request:
## Which issue does this PR close?
Related to Epic: https://github.com/apache/datafusion-comet/issues/1042
`arrays_overlap`: `select arrays_overlap(array('hello', '-', 'world'), array('hello'))` => `true`
DataFusion' s array_has_any has same behavior with Spark 's arrays_overlap function
`Spark:` https://docs.databricks.com/en/sql/language-manual/functions/arrays_overlap.html
`DataFusion:` https://datafusion.apache.org/user-guide/sql/scalar_functions.html#array-has-any

## Rationale for this change
Defined under Epic: https://github.com/apache/datafusion-comet/issues/1042

## What changes are included in this PR?
`planner.rs:` Maps Spark 's `arrays_overlap` function to DataFusion `array_has_any` physical expression from Spark physical expression with return type: DataType::Boolean,
`expr.proto:` arrays_overlap message has been added,
`QueryPlanSerde.scala:` arrays_overlap pattern matching case has been added,
`CometExpressionSuite.scala:` A new UT has been added for arrays_overlap function.

## How are these changes tested?
A new UT has been added.

andygrove commented on spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:
```diff
@@ -2300,6 +2300,12 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
             expr.children(1),
             inputs,
             (builder, binaryExpr) => builder.setArrayAppend(binaryExpr))
+        case ArraysOverlap(leftArrayExpr, rightArrayExpr) =>
```
Do we support this for all input data types?

andygrove replied:
Because the PR only contains basic tests, could you add a check to enable this expression only if `CometConf.COMET_CAST_ALLOW_INCOMPATIBLE` is enabled? We can remove this check in a future PR that adds comprehensive tests and demonstrates that we have Spark-compatible behavior for all supported data types.`

erenavsarogullari replied:
Yes, makes sense and `COMET_CAST_ALLOW_INCOMPATIBLE` check has been added.

andygrove commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -2568,4 +2568,21 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
     }
   }
+
+  test("arrays_overlap") {
```
Could you add tests where one or both input expressions evaluate to null on some rows? This can be achieved with a CASE expression as in some of the other array tests

kazuyukitanimura replied:
+1 for more coverage

erenavsarogullari replied:
Case Expression based test case has been added.

andygrove commented:
@erenavsarogullari could you rebase this PR when you have time? We can go ahead and merge this and then improve tests as part of https://github.com/apache/datafusion-comet/issues/1269

kazuyukitanimura commented on spark/src/main/scala/org/apache/comet/serde/QueryPlanSerde.scala:
```diff
@@ -2300,6 +2300,12 @@ object QueryPlanSerde extends Logging with ShimQueryPlanSerde with CometExprShim
             expr.children(1),
             inputs,
             (builder, binaryExpr) => builder.setArrayAppend(binaryExpr))
+        case ArraysOverlap(leftArrayExpr, rightArrayExpr) =>
```
Would you add a flag to enable/disable?

andygrove replied:
We could use the existing `COMET_CAST_ALLOW_INCOMPATIBLE` config to enable this function until we have comprehensive testing in place.

erenavsarogullari commented:
> @erenavsarogullari could you rebase this PR when you have time? We can go ahead and merge this and then improve tests as part of #1269

Thanks @andygrove and @kazuyukitanimura for the review. PR has been rebased and addressed the comments.

erenavsarogullari commented:
FYI, Currently, DataFusion does not support `arrays_overlap` function but supports same functionality by `array_has_any` function. Other query engines like `Snowflake`, `Presto` support `arrays_overlap` function. In terms of this, also created DataFusion PR: https://github.com/apache/datafusion/pull/14217

andygrove reviewed (approved):
Thanks @erenavsarogullari. LGTM.

andygrove commented on spark/src/test/scala/org/apache/comet/CometExpressionSuite.scala:
```diff
@@ -2701,4 +2701,25 @@ class CometExpressionSuite extends CometTestBase with AdaptiveSparkPlanHelper {
+          checkSparkAnswerAndOperator(sql(
+            "SELECT arrays_overlap(array('a', null), array('b', null)) from t1 where _1 is not null"))
+          checkSparkAnswerAndOperator(spark.sql(
+            "SELECT arrays_overlap((CASE WHEN _2 =_3 THEN array(_6, _7) END), array(_6, _7)) FROM t1"));
```
I think `_2 = _3` may always be true. The problem with `makeParquetFileAllTypes` is that every for each row, each column contains the same integer value cast to the column's type, so it is not ideal for tests like this. We can improve as part of https://github.com/apache/datafusion-comet/issues/1269
